v1.52.0 Failed to connect to Tailscale service

### What is the issue?

---------------------------
Tailscale Error
---------------------------
Failed to connect to Tailscale service.



Version 1.52.0 (t3c2ff1e4a-gfccfad18e)



LogID: d7215ddcb08bb19ce0ab4ea8bcfa79f9bee1ba3764701c131f3cd9fc888038c9

(Ctrl-C to copy this text)


---------------------------
确定   
---------------------------


### Steps to reproduce

Normally start tailscale service (or after reboot)

### Are there any recent changes that introduced the issue?

After updated from 1.51.0 to 1.52.0, this bug occurs.

### OS

Windows

### OS version

Windows 10

### Tailscale version

1.52.0

### Other software

zerotier
clash for windows

### Bug report

d7215ddcb08bb19ce0ab4ea8bcfa79f9bee1ba3764701c131f3cd9fc888038c9

> @llllllllqq @kegns @GD-Slime could you open a `cmd.exe` shell and run `tailscaled.exe` and paste the output here? We're looking for whether it is similar to [#10023 (comment)](https://github.com/tailscale/tailscale/issues/10023#issuecomment-1786504326)

logtail started
local disk logdir: C:\ProgramData\Tailscale\Logs
Program starting: v1.52.0-t3c2ff1e4a-gfccfad18e, Go 1.21.3: []string{"C:\\Program Files\\Tailscale\\tailscaled.EXE"}
cleaning log files; number of files 51 > 50
cleaned log file C:\ProgramData\Tailscale\Logs\tailscale-service-20230916T144221-1694846541.txt (size 227329); new bytes=10937749, files=50
LogID: 7e45a7fd5ddac2e918fa0b68938649959e47f1f21d8fd8d70f5d1961602136a1
logpolicy: using dir C:\ProgramData\Tailscale
wgengine.NewUserspaceEngine(tun "Tailscale") ...
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x20 pc=0x5f0605]

goroutine 22 [running]:
golang.org/x/sys/windows.UTF16PtrToString(0x0?)
        golang.org/x/sys@v0.13.0/windows/syscall_windows.go:119 +0x25
golang.zx2c4.com/wintun.logMessage(0x0?, 0xda985fd960, 0x0?)
        golang.zx2c4.com/wintun@v0.0.0-20230126152724-0fa3db229ce2/wintun.go:51 +0xff
syscall.SyscallN(0x7ffea9843fb0?, {0xc0002b1898?, 0x3?, 0xc0002b18b0?})
        runtime/syscall_windows.go:544 +0x107
syscall.Syscall(0x12c60e0?, 0xc000032aa0?, 0xc0002b1978?, 0x57df85?, 0xc2108a?)
        runtime/syscall_windows.go:482 +0x35
golang.zx2c4.com/wintun.CreateAdapter({0xc12ea4?, 0x45d964b800?}, {0xc12ea4, 0x9}, 0xc00000a760)
        golang.zx2c4.com/wintun@v0.0.0-20230126152724-0fa3db229ce2/wintun.go:95 +0xf9
github.com/tailscale/wireguard-go/tun.CreateTUNWithRequestedGUID({0xc12ea4, 0x9}, 0xd6ea90?, 0x500)
        github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/tun/tun_windows.go:69 +0x45
github.com/tailscale/wireguard-go/tun.CreateTUN(...)
        github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/tun/tun_windows.go:63
tailscale.com/net/tstun.New(0xc0000a60a0, {0xc12ea4, 0x9})
        tailscale.com@v1.52.0/net/tstun/tun.go:47 +0x85
main.tstunNewWithWindowsRetries(0xc0000a60a0, {0xc12ea4, 0x9})
        tailscale.com@v1.52.0/cmd/tailscaled/tailscaled_windows.go:93 +0x119
main.tryEngine(0xc3b35e?, 0xc000132d80, {0xc12ea4, 0x9})
        tailscale.com@v1.52.0/cmd/tailscaled/tailscaled.go:657 +0x2ef
main.createEngine(0xc0000a60a0, 0xbd1de0?)
        tailscale.com@v1.52.0/cmd/tailscaled/tailscaled.go:591 +0x19d
main.getLocalBackend({0x0?, 0x0?}, 0xc0000a60a0, {0x7e, 0x45, 0xa7, 0xfd, 0x5d, 0xda, 0xc2, ...}, ...)
        tailscale.com@v1.52.0/cmd/tailscaled/tailscaled.go:498 +0x11a
main.startIPNServer.func2()
        tailscale.com@v1.52.0/cmd/tailscaled/tailscaled.go:464 +0x259
created by main.startIPNServer in goroutine 1
        tailscale.com@v1.52.0/cmd/tailscaled/tailscaled.go:452 +0x44a

--- a/cmd/tailscaled/tailscaled_windows.go
+++ b/cmd/tailscaled/tailscaled_windows.go
@@ -30,6 +30,7 @@ import (
 	"os"
 	"os/exec"
 	"os/signal"
+	"path/filepath"
 	"sync"
 	"syscall"
 	"time"
@@ -299,6 +300,14 @@ func beWindowsSubprocess() bool {
 		}
 	}()
 
+	// Pre-load wintun.dll using a fully-qualified path so that wintun-go
+	// loads our copy and not some (possibly outdated) copy dropped in system32.
+	// (OSS Issue #10023)
+	fqWintunPath := fullyQualifiedWintunPath(log.Printf)
+	if _, err := windows.LoadDLL(fqWintunPath); err != nil {
+		log.Printf("Error pre-loading \"%s\": %v", fqWintunPath, err)
+	}
+
 	sys := new(tsd.System)
 	netMon, err := netmon.New(log.Printf)
 	if err != nil {
@@ -507,7 +516,7 @@ func babysitProc(ctx context.Context, args []string, logf logger.Logf) {
 }
 
 func uninstallWinTun(logf logger.Logf) {
-	dll := windows.NewLazyDLL("wintun.dll")
+	dll := windows.NewLazyDLL(fullyQualifiedWintunPath(logf))
 	if err := dll.Load(); err != nil {
 		logf("Cannot load wintun.dll for uninstall: %v", err)
 		return
@@ -516,4 +525,17 @@ func uninstallWinTun(logf logger.Logf) {
 	logf("Removing wintun driver...")
 	err := wintun.Uninstall()
 	logf("Uninstall: %v", err)
+}
+
+func fullyQualifiedWintunPath(logf logger.Logf) string {
+	var dir string
+	var buf [windows.MAX_PATH]uint16
+	length := uint32(len(buf))
+	if err := windows.QueryFullProcessImageName(windows.CurrentProcess(), 0, &buf[0], &length); err != nil {
+		logf("QueryFullProcessImageName failed: %v", err)
+	} else {
+		dir = filepath.Dir(windows.UTF16ToString(buf[:length]))
+	}
+
+	return filepath.Join(dir, "wintun.dll")
 }

